generate_devices_list.py

```diff
@@ -23,104 +22,32 @@ def load_and_parse_json(file_name):
-  if not ("top" in outline["insets"]) or not isinstance(outline["insets"]["top"], numbers.Number) or outline["insets"]["top"] < 0:
-    raise_type_error(file_name, prefix + "/outline/insets/top", "number")
-
-def parse_modes(modes, file_name, prefix, rebase_path):
-  for (index, mode) in enumerate(modes):
-    i = str(index)
-    if not isinstance(mode, dict):
-      raise_type_error(file_name, prefix + "[" + i + "]", "object")
-    if not ("title" in mode) or not isinstance(mode["title"], basestring):
-      raise_type_error(file_name, prefix + "[" + i + "]/title", "string")
-    if not ("orientation" in mode) or not isinstance(mode["orientation"], basestring):
-      raise_type_error(file_name, prefix + "[" + i + "]/orientation", "string")
-    if mode["orientation"] != "vertical" and mode["orientation"] != "horizontal":
-      raise Exception('ERROR: "' + prefix + '[' + i + ']/orientation" must be one of the following: [horizontal, vertical] (' + file_name + ')')
-
-    if not ("images" in mode) or not isinstance(mode["images"], (list, tuple)):
-      raise_type_error(file_name, prefix + "[" + i + "]/images", "array")
-    parse_and_rebase_images(mode["images"], file_name, prefix + "[" + i + "]/images", rebase_path)
-
-    if not ("page-rect" in mode) or not isinstance(mode["page-rect"], dict):
-      raise_type_error(file_name, prefix + "[" + i + "]/page-rect", "object")
-    if not ("left" in mode["page-rect"]) or not isinstance(mode["page-rect"]["left"], numbers.Number) or mode["page-rect"]["left"] < 0:
-      raise_type_error(file_name, prefix + "[" + i + "]/page-rect/left", "object")
-    if not ("top" in mode["page-rect"]) or not isinstance(mode["page-rect"]["top"], numbers.Number) or mode["page-rect"]["top"] < 0:
-      raise_type_error(file_name, prefix + "[" + i + "]/page-rect/top", "object")
-    if not ("width" in mode["page-rect"]) or not isinstance(mode["page-rect"]["width"], numbers.Number) or mode["page-rect"]["width"] < 0:
-      raise_type_error(file_name, prefix + "[" + i + "]/page-rect/width", "object")
-    if not ("height" in mode["page-rect"]) or not isinstance(mode["page-rect"]["height"], numbers.Number) or mode["page-rect"]["height"] < 0:
-      raise_type_error(file_name, prefix + "[" + i + "]/page-rect/height", "object")
-
-def parse_device_json(file_name, rebase_path):
-  json = load_and_parse_json(file_name)
 
-  if not ("title" in json) or not isinstance(json["title"], basestring):
-    raise_type_error(file_name, "title", "string")
+  parse_and_rebase_images(d["outline"]["images"], file_name)
 
-  if not ("type" in json) or not isinstance(json["type"], basestring):
-    raise_type_error(file_name, "type", "string")
-  if not (json["type"] in POSSIBLE_TYPES):
-    raise Exception('ERROR: "type" must be one of the following: [' + ", ".join(POSSIBLE_TYPES) + '] (' + file_name + ')')
+def parse_modes(modes, file_name):
+  for mode in modes:
+    parse_and_rebase_images(mode["images"], file_name)
```

Comment:
file_name and rebase_path are used one instead of another. You should only pass around rebase_path, and drop file_name and prefix parameters.